src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+"""
+## A Module for Optimization with Various Regularizations.
+
+This module provides a series of classes that facilitate the optimization of models
+with different types of regularizations. Each solver class in this module interfaces
+with various optimization methods, and they can be applied depending on the model's requirements.
+"""
+import abc
+import inspect
+from typing import Callable, List, Optional, Tuple, Union
+
+import jax.numpy as jnp
+import jaxopt
+from numpy.typing import NDArray
+
+from .base_class import _Base
+from .proximal_operator import prox_group_lasso
+
+__all__ = ["UnRegularizedSolver", "RidgeSolver", "LassoSolver", "GroupLassoSolver"]
+
+
+def __dir__() -> list[str]:
+    return __all__
+
+
+class Solver(_Base, abc.ABC):
+    """
+    Abstract base class for optimization solvers.
+
+    This class is designed to provide a consistent interface for optimization solvers,
+    enabling users to easily switch between different solvers and ensure compatibility
+    with various loss functions and regularization schemes.
+
+    Attributes
+    ----------
+    allowed_optimizers :
+        List of optimizer names that are allowed for use with this solver.
+    solver_name :
+        Name of the solver being used.
+    solver_kwargs :
+        Additional keyword arguments to be passed to the solver during instantiation.
+
+    Methods
+    -------
+    instantiate_solver(loss) :
+        Abstract method to instantiate a solver with a given loss function.
+    get_runner(solver_kwargs, run_kwargs) :
+        Get the solver runner with provided arguments.
```

Comment:
I'm against documenting methods like this: hard to keep up to date, redundant with the autodoc'ing

Reply:
removed them